src/main/kotlin/uk/gov/dluhc/printapi/database/repository/AnonymousElectorDocumentSummarySpecificationBuilder.kt

```diff
@@ -0,0 +1,75 @@
+import javax.persistence.criteria.Root
+
+@Component
+class AnonymousElectorDocumentSummarySpecificationBuilder {
+
+    companion object {
+        private const val GSS_CODE: String = "gssCode"
+        private const val APPLICATION_REFERENCE: String = "applicationReference"
+        private const val SANITIZED_SURNAME: String = "sanitizedSurname"
+    }
+
+    fun buildSpecification(
+        gssCodes: List<String>,
+        criteria: AnonymousSearchCriteriaDto
+    ): Specification<AnonymousElectorDocumentSummary> {
+        return buildSpecificationForGssCodes(gssCodes)
+            .and(buildSpecificationForSearchBy(criteria.searchBy, criteria.searchValue))
+    }
+
+    private fun buildSpecificationForGssCodes(gssCodes: List<String>) =
+        hasGssCodeIn(gssCodes)
+
+    private fun buildSpecificationForSearchBy(
+        searchBy: AedSearchBy?,
+        searchValue: String?
+    ): Specification<AnonymousElectorDocumentSummary>? {
+        if (searchBy == null || searchValue == null) {
+            return null
+        }
+
+        return when (searchBy) {
+            AedSearchBy.SURNAME -> hasSanitizedSurname(sanitizeSurname(searchValue))
+            AedSearchBy.APPLICATION_REFERENCE -> hasApplicationReference(sanitizeApplicationReference(searchValue))
+        }
+    }
+
+    private fun hasGssCodeIn(gssCodes: List<String>): Specification<AnonymousElectorDocumentSummary> {
+        return Specification { root: Root<AnonymousElectorDocumentSummary>, _: CriteriaQuery<*>?, criteriaBuilder: CriteriaBuilder ->
+            val query: CriteriaQuery<AnonymousElectorDocumentSummary> = criteriaBuilder.createQuery(
+                AnonymousElectorDocumentSummary::class.java
+            )
+            query.select(root).where(root.get<Any>(GSS_CODE).`in`(gssCodes)).restriction
+        }
+    }
+
+    private fun hasApplicationReference(applicationReference: String): Specification<AnonymousElectorDocumentSummary> {
+        return Specification<AnonymousElectorDocumentSummary> { root: Root<AnonymousElectorDocumentSummary?>, _: CriteriaQuery<*>?, criteriaBuilder: CriteriaBuilder ->
+            criteriaBuilder.equal(root.get<Any>(APPLICATION_REFERENCE), applicationReference)
+        }
+    }
+
+    private fun hasSanitizedSurname(sanitizedSurname: String): Specification<AnonymousElectorDocumentSummary> {
+        return Specification<AnonymousElectorDocumentSummary> { root: Root<AnonymousElectorDocumentSummary?>, _: CriteriaQuery<*>?, criteriaBuilder: CriteriaBuilder ->
+            criteriaBuilder.equal(root.get<Any>(SANITIZED_SURNAME), sanitizedSurname)
+        }
+    }
+
+    private fun sanitizeSurname(surname: String): String {
+        return surname.uppercase().replace(Regex("-"), " ").replace(Regex("'"), "").replace(Regex("[ ]{2,}"), " ")
+            .trim()
+    }
+
+    private fun sanitizeApplicationReference(applicationReference: String): String {
+        return applicationReference.uppercase().replace(Regex("[ ]+"), "")
```

Comment:
Do we always persist the `applicationReference` in uppercase in print-api db? 
If not, then we might need to sanitize the `applicationReference` in view for exact match

Reply:
Thats a good question. I assumed we do, but I'll check 👍

Reply:
I've had a look, and whilst we don't explicitly uppercase the ref number when storing in the AED table, or in the view DDL, I think it's a fair assumption that it will always be uppercase:

* The application reference value is not from user input. AED's are created in this DB via an SQS or REST message from VCA. Its not a REST API from the UI (so therefore the user is not in control of the app ref number)
* The application reference comes "direct" from VCA's database
* The application reference comes either from IER's original CreateVoterCardApplication message, or VCA's internal code in the case of an AED or paper application

I think it's a safe assumption that the value in `print-api`'s database will always be uppercase 👍

Reply:
That sounds like a safe assumption in terms of how the system works in live, but what about Josh and postman?! 😉

Reply:
🤐